### DISCUSSION TITLE
Calibrating multiple motors (hardware stop) using the MC4-PLUS board

### DISCUSSION THREAD

**Discussion by plinioMoreno** [robotology/community#608] [Q&A] [answered]
We are creating the configuration files for our robot Vizzy, which will use the ETH-based boards (the repository of the files is [here](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files)). We are running into issues when calibrating two DC motors at the same time. When running the calibration at the same time (see [config files](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/6e4064965525f1355b418d4f0861fe1b0ee7c058/calibrators/head-calib.xml)), the pwm is commanded and the motors start to move to reach the hard stop, but the motors do not move to the home position, and the message we get in the console is that the calibration is finished and worked properly. 

To figure out if the problem is from the files, we run the calibration independently for each motor using these [config files](https://github.com/vislab-tecnico-lisboa/eth-boards-setups/blob/416bd9f2608899d2278ca7e66efbdac7b59edda1/single-ETH-1-motor-test-vislab/calibrators/knee-calib.xml), and in this case the calibration works as expected.

Since the parameters of the calibration for both tests were the same, we wonder how can we figure out the problem. Any help is very appreciated.

cc @marcoaccame @ale-git @Ryo-Naohara

**Comment by marcoaccame** [comment, marked as answer]
Hi @plinioMoreno, 
we had a look at the xml files and we surely have some basic suggestions. For instance, the file [**head-eb05-j0_1-mec.xml**](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/main/hardware/mechanicals/head-eb05-j0_1-mec.xml ) seems wrong in the the `COUPLING` and `JOINTSET_CFG` groups.

To give more focused suggestions, however, we also need to know some more about the mechanical setup.

We can start from one point: the step-by-step first movement of joints 0 and 1. 

### Step by step movement of joints 0 and 1.  

To debug your setup you can begin to calibrate and move joint 0 and 1 considering them independent joints. That is what you have probably done using the other [file](https://github.com/vislab-tecnico-lisboa/eth-boards-setups/blob/416bd9f2608899d2278ca7e66efbdac7b59edda1/single-ETH-1-motor-test-vislab/calibrators/knee-calib.xml) you cited.  After you are sure that the motors move the joints in the way you expect (in direction and angle and PID) you go forward. 

At this point if the joints are independent, the task is over, otherwise if the joints are coupled you must add the coupling rule specific of your mechanical setup.

Let's see how to do that in steps:

1. Move joint 0 and 1 (or the motors) in independent mode.

   - You should change [**head-eb05-j0_1-mec.xml**](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/6e4064965525f1355b418d4f0861fe1b0ee7c058/hardware/mechanicals/head-eb05-j0_1-mec.xml) so that the `COUPLING` and `JOINTSET_CFG` groups are:

   ```xml
        <group name="COUPLINGS">
   		<!-- marco.accame: use identity matric for independent joints / motors  
             -->
           <param name="matrixJ2M">  
               1.00    0.00    0.00    0.00
               0.00    1.00    0.00    0.00
               0.00    0.00    1.00    0.00
               0.00    0.00    0.00    1.00 
           </param>
   
           <param name="matrixM2J">  
               1.00    0.00    0.00    0.00
               0.00    1.00    0.00    0.00
               0.00    0.00    1.00    0.00
               0.00    0.00    0.00    1.00 
           </param>
            
           <param name ="matrixE2J"> 
               1.00    0.00    0.00    0.00    0.00    0.00
               0.00    1.00    0.00    0.00    0.00    0.00
               0.00    0.00    1.00    0.00    0.00    0.00
               0.00    0.00    0.00    1.00    0.00    0.00
           </param>
   
       </group>
   
       <group name="JOINTSET_CFG"> 
   		<!-- marco.accame: use two jointsets for independent joints 
             -->        
           <param name= "numberofsets"> 2 </param>
           <group name="JOINTSET_0">
               <param name="listofjoints">  0             </param>
               <param name="constraint">    none          </param>
               <param name="param1">        0             </param>
               <param name="param2">        0             </param>
           </group>
           <group name="JOINTSET_1">
               <param name="listofjoints">  1             </param>
               <param name="constraint">    none          </param>
               <param name="param1">        0             </param>
               <param name="param2">        0             </param>
           </group>        
       </group>
   
   ```

   - keep or modify the following files according to the properties of your setup: [**head-eb05-j0_1-mc.xml**]( https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/6e4064965525f1355b418d4f0861fe1b0ee7c058/hardware/motorControl/head-eb05-j0_1-mc.xml), [**head-eb05-j0_1-mc_service.xml**](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/6e4064965525f1355b418d4f0861fe1b0ee7c058/hardware/motorControl/head-eb05-j0_1-mc_service.xml), the `GENERAL` and `LIMITS` groups of [**head-eb05-j0_1-mec.xml**](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/6e4064965525f1355b418d4f0861fe1b0ee7c058/hardware/mechanicals/head-eb05-j0_1-mec.xml).

   - change [**head-calib.xml**](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/6e4064965525f1355b418d4f0861fe1b0ee7c058/calibrators/head-calib.xml) so that you have:

     ```xml
     <param name="CALIB_ORDER"> (0) (1) (2) (3) (4) </param> 
     ```

   - edit the **[general.xml](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/6e4064965525f1355b418d4f0861fe1b0ee7c058/general.xml)** so that you skip calibration:

     ```xml
     <param name="skipCalibration">    true  </param>
     ```

   - launch `yarprobotinterface` and then `yarpmotorgui --calib`

   - attempt calibration on joint 0 and then on joint 1.

   - if everything is OK you can also revert skipCalibration to be false again and restart `yarprobotinterface`.

2. move joint 0 and 1 (or the motors) in coupled mode. See in here an example of coupled joints: [the pitch and roll of the neck of iCub](https://github.com/robotology/robots-configuration/blob/master/iCubGenova02/hardware/mechanicals/head-eb20-j0_1-mec.xml).
   - find out the coupling rule between joint 0 and joint1 and add it inside [**head-eb05-j0_1-mec.xml**](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/6e4064965525f1355b418d4f0861fe1b0ee7c058/hardware/mechanicals/head-eb05-j0_1-mec.xml). 
   
   - in teh same file add one jointset only w/  both the joints. 
   - calibrate teh two joints together:
      ```xml
     <param name="CALIB_ORDER"> (0 1) (2) (3) (4) </param> 
     ```

**Reply by plinioMoreno** [reply]
Hi @marcoaccame,

Thanks for the informative reply. 

Let me summarize the mechanicals of our robot: Most of the joints (head, torso and arms) are independent joints. The only coupled joints are the two eyes, similarly to iCub (version and vergence coupled in control software) for which I should follow [head-eb21-j2_5-mec.xml](https://github.com/robotology/robots-configuration/blob/master/iCubGenova02/hardware/mechanicals/head-eb21-j2_5-mec.xml#L41), is that right?

The following picture and table show more details about the mechanicals, where we have four serial chains of revolute joints: left arm, right, arm, left eye and right eye:

![imagem](https://user-images.githubusercontent.com/3798890/218408305-36fafdf3-9101-44fa-80ee-89f9b7163e38.png)
![imagem](https://user-images.githubusercontent.com/3798890/218408601-ef35212b-8b20-472d-aa55-0cd47be993ad.png)

We were able to calibrate sucessfully the two joints! We changed the `COUPLING` and `JOINTSET_CFG` as you explained on [our file](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/d056a9eb80f64392344373ce4ce610bc43d25ef2/hardware/mechanicals/head-eb05-j0_1-mec.xml#L35):

```xml
    <group name="COUPLINGS">
		<!-- Check if this is right  -->
        <param name="matrixJ2M">  
            1.00    0.00    0.00    0.00
            0.00    1.00    0.00    0.00
            0.00    0.00    1.00    0.00
            0.00    0.00    0.00    1.00 
        </param>
		<!-- Check if this is right  -->
        <param name="matrixM2J">  
            1.00    0.00    0.00    0.00
            0.00    1.00    0.00    0.00
            0.00    0.00    1.00    0.00
            0.00    0.00    0.00    1.00 
        </param>
		<!-- Check if this is right  -->
        <param name ="matrixE2J"> 
            1.00    0.00    0.00    0.00    0.00    0.00
            0.00    1.00    0.00    0.00    0.00    0.00
            0.00    0.00    1.00    0.00    0.00    0.00
            0.00    0.00    0.00    1.00    0.00    0.00
        </param>

    </group>

    <group name="JOINTSET_CFG"> <!-- To be deleted  -->
        <param name= "numberofsets"> 2 </param>
        <group name="JOINTSET_0">
            <param name="listofjoints">  0           </param>
            <param name="constraint">    none          </param>
            <param name="param1">        0             </param>
            <param name="param2">        0             </param>
        </group>
        <group name="JOINTSET_1">
            <param name="listofjoints">  1           </param>
            <param name="constraint">    none          </param>
            <param name="param1">        0             </param>
            <param name="param2">        0             </param>
        </group>
    </group>
```
We are able to run the calibration either in sequence
```xml
<param name="CALIB_ORDER"> (0) (1) </param>
```
or at the same time on the two independent motors
```xml
<param name="CALIB_ORDER"> (0 1) </param>
```
We noticed that the `startupPosThreshold` had to be updated to 3.5 degrees for the calibration to finish properly ([file](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/d056a9eb80f64392344373ce4ce610bc43d25ef2/calibrators/head-calib.xml#L26)) Do you know how to reduce this threshold?

**Reply by pattacini** [reply]
Hi @plinioMoreno 

>I should follow [head-eb21-j2_5-mec.xml](https://github.com/robotology/robots-configuration/blob/master/iCubGenova02/hardware/mechanicals/head-eb21-j2_5-mec.xml#L41), is that right?

Right 👍🏻 
More precisely, both the matrices: [`matrixJ2M` and `matrixM2J`][1].

>We noticed that the startupPosThreshold had to be updated to 3.5 degrees for the calibration to finish properly ([file](https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files/blob/d056a9eb80f64392344373ce4ce610bc43d25ef2/calibrators/head-calib.xml#L26)) Do you know how to reduce this threshold?

Maybe, @ale-git can give you some heads-up on the meaning and functionality of `startupPosThreshold`.

[1]: https://github.com/robotology/robots-configuration/blob/master/iCubGenova02/hardware/mechanicals/head-eb21-j2_5-mec.xml#L34-L46

**Reply by plinioMoreno** [reply]
Hi @pattacini,

I saw that the matrices [matrixJ2M and matrixM2J](https://github.com/robotology/robots-configuration/blob/master/iCubGenova02/hardware/mechanicals/head-eb21-j2_5-mec.xml#L34-L46) consider that the first joint is the right eye and the second one is the left eye [as per commented here](https://github.com/robotology/robots-configuration/blob/master/iCubGenova02/calibrators/head-calib.xml#L20). How do those J2M and M2J matrices look like if the joints are in reversed order (first left eye and then right eye)?

**Reply by pattacini** [reply]
Hi @plinioMoreno,

The transformation between version|vergence to/from right|left eyes is computed according to https://icub-tech-iit.github.io/documentation/icub_kinematics/icub-vergence-version/icub-vergence-version.

The comment you referred to is a bit misleading as joints are only version and vergence and not the eyes, although the order of the eyes is correct.

If you invert the order of the eyes, then you have:
```xml
    <param name ="matrixM2J"> 
        1.000   0.000   0.000   0.000
        0.000   1.000   0.000   0.000
        0.000   0.000   0.500   0.500
        0.000   0.000   0.500  -0.500   
    </param>
```

`matrixJ2M` is merely the inverse of `matrixM2J`.

Also, be careful, as the numbers in the matrices should be fixed as per https://github.com/robotology/robots-configuration/issues/228.

**Comment by pattacini** [comment]
Hi @plinioMoreno 

A somehow related question.

Since it seems you've been hosting XML files for `Vizzy` in https://github.com/vislab-tecnico-lisboa/vizzy-eth-config-files, what about the files currently available at https://github.com/robotology/robots-configuration/tree/devel/vizzy?

Are those files still employed? May I clean them up?

**Reply by plinioMoreno** [reply]
Hi @pattacini,

Vizzy still has its CAN-based control boards functional, so we are still using the files at [robots-configuration](https://github.com/robotology/robots-configuration/tree/devel/vizzy). The ETH setup we had is for testing purposes before mounting the boards and when we have the new configuration working, we will put the new files at the robots-configuration repository.

I'll let you know when the files are ready for clean up.

**Reply by pattacini** [reply]
Ok @plinioMoreno, thanks for the clarification 👍🏻 

My only doubt was whether you would have started using your own repos. This is not the case, so please continue relying on `robots-configuration`. You could also deal with the possible clean up as I don't maintain robot-specific files inside the folders.

**Comment by ale-git** [comment]
The precision of the startup movement could depend on two factors: 
- the precision of the control itself, and thus you should check the tuning of the position PID
- if reaching the startupPosition requires high torque maybe the startupMaxPwm is too low?
